dia-batching-server/src/api/coingecko.rs

```diff
@@ -0,0 +1,310 @@
+	fn convert_to_coingecko_id(asset: &AssetSpecifier) -> Option<String> {
+		match (asset.blockchain.as_str(), asset.symbol.as_str()) {
+			("Pendulum", "PEN") => Some("pendulum".to_string()),
+			("Polkadot", "DOT") => Some("polkadot".to_string()),
+			("Kusama", "KSM") => Some("kusama".to_string()),
+			("Astar", "ASTR") => Some("astar".to_string()),
+			("Bifrost", "BNC") => Some("bifrost-native-coin".to_string()),
+			("Bifrost", "vDOT") => Some("voucher-dot".to_string()),
+			("HydraDX", "HDX") => Some("hydration".to_string()),
+			("Moonbeam", "GLMR") => Some("moonbeam".to_string()),
+			("Polkadex", "PDEX") => Some("polkadex".to_string()),
+			("Stellar", "XLM") => Some("stellar".to_string()),
+			_ => None,
+		}
+	}
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct SimplePing {
+	pub gecko_says: String,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct CoingeckoPrice {
+	pub usd: Decimal,
+	pub usd_market_cap: Option<Decimal>,
+	pub usd_24h_vol: Option<Decimal>,
+	pub usd_24h_change: Option<Decimal>,
+	pub last_updated_at: u64,
+}
+
+/// CoinGecko network client
+pub struct CoingeckoClient {
+	host: String,
+	api_key: String,
+}
+
+impl CoingeckoClient {
+	pub fn new(host: String, api_key: String) -> Self {
+		CoingeckoClient { host, api_key }
+	}
+
+	async fn get<R: DeserializeOwned>(&self, endpoint: &str) -> Result<R, CoingeckoError> {
+		let mut headers = reqwest::header::HeaderMap::new();
+		headers.insert("accept", reqwest::header::HeaderValue::from_static("application/json"));
+
+		// We supply a different header for the demo API
+		let mut api_key = reqwest::header::HeaderValue::from_str(self.api_key.as_str())
+			.map_err(|e| CoingeckoError(format!("Could not set API key header value: {}", e)))?;
+		api_key.set_sensitive(true);
+		if self.host.contains("pro-api") {
+			headers.insert("x-cg-pro-api-key", api_key);
+		} else {
+			headers.insert("x-cg-demo-api-key", api_key);
+		}
+
+		let client = reqwest::Client::builder()
+			.default_headers(headers)
+			.build()
+			.map_err(|e| CoingeckoError(e.to_string()))?;
+
+		let url = reqwest::Url::parse(
+			format!("{host}/{ep}", host = self.host.as_str(), ep = endpoint).as_str(),
+		)
+		.expect("Invalid URL");
+
+		let response = client.get(url).send().await;
+
+		match response {
+			Ok(response) => {
+				if !response.status().is_success() {
+					let result = response.text().await;
+					let err_msg = format!(
+						"CoinGecko API error: {}",
+						result.unwrap_or("Unknown".to_string()).trim()
+					);
+					Err(CoingeckoError(err_msg))
+				} else {
+					let result = response.json().await;
+					result.map_err(|e| {
+						CoingeckoError(format!("Could not decode CoinGecko response: {}", e))
+					})
+				}
+			},
+			Err(e) => Err(CoingeckoError(e.to_string())),
+		}
+	}
+
+	/// Check API server status
+	#[allow(dead_code)]
+	pub async fn ping(&self) -> Result<SimplePing, CoingeckoError> {
```

Comment:
It's nice that they have this endpoint for checking the server status, but how often do we plan to call it and what would be the "retry logic" on our side in case their server is down?

Reply:
I added this at first because I thought it would come in handy for testing if the server is down or not but then I realized that it might not be necessary to have that and we can just try to get a response from the usual endpoint and log the error in case something fails. There is nothing to retry or fall back to anyway.